packages/contracts-bedrock/scripts/checks/common/util.go

```diff
@@ -0,0 +1,120 @@
+package common
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"github.com/bmatcuk/doublestar/v4"
+)
+
+type ErrorReporter struct {
+	hasErr atomic.Bool
+	outMtx sync.Mutex
+}
+
+func NewErrorReporter() *ErrorReporter {
+	return &ErrorReporter{}
+}
+
+func (e *ErrorReporter) Fail(msg string, args ...any) {
+	e.outMtx.Lock()
+	// Useful for suppressing error reporting in tests
+	if os.Getenv("SUPPRESS_ERROR_REPORTER") == "" {
+		_, _ = fmt.Fprintf(os.Stderr, "❌  "+msg+"\n", args...)
+	}
+	e.outMtx.Unlock()
+	e.hasErr.Store(true)
+}
+
+func (e *ErrorReporter) HasError() bool {
+	return e.hasErr.Load()
+}
+
+type FileProcessor func(path string) error
+
+func ProcessFiles(files map[string]string, processor FileProcessor) error {
+	reporter := NewErrorReporter()
+	sem := make(chan struct{}, runtime.NumCPU())
```

Comment:
Highly recommend the `errgroup` package for this, removes needing to manually implement concurrency primitives and has a nice API. Example: https://pkg.go.dev/golang.org/x/sync/errgroup#example-Group-Parallel

Reply:
Ok I'll look into errgroup instead

Reply:
Using errgroup now.